src/hotspot/share/gc/parallel/psPromotionManager.inline.hpp

```diff
@@ -153,6 +153,80 @@ inline oop PSPromotionManager::copy_to_survivor_space(oop o) {
   }
 }
 
+inline HeapWord* PSPromotionManager::allocate_in_young_gen(Klass* klass,
+                                                           size_t obj_size,
+                                                           uint age) {
+  HeapWord* result = _young_lab.allocate(obj_size);
+  if (result != nullptr) {
+    return result;
+  }
+  if (_young_gen_is_full) {
+    return nullptr;
+  }
+  // Do we allocate directly, or flush and refill?
+  if (obj_size > (YoungPLABSize / 2)) {
+    // Allocate this object directly
+    result = young_space()->cas_allocate(obj_size);
+    promotion_trace_event(cast_to_oop(result), klass, obj_size, age, false, nullptr);
+  } else {
+    // Flush and fill
+    _young_lab.flush();
+
+    HeapWord* lab_base = young_space()->cas_allocate(YoungPLABSize);
+    if (lab_base != nullptr) {
+      _young_lab.initialize(MemRegion(lab_base, YoungPLABSize));
+      // Try the young lab allocation again.
+      result = _young_lab.allocate(obj_size);
+      promotion_trace_event(cast_to_oop(result), klass, obj_size, age, false, &_young_lab);
+    } else {
+      _young_gen_is_full = true;
+    }
+  }
+  if (result == nullptr && !_young_gen_is_full && !_young_gen_has_alloc_failure) {
```

Comment:
Do you need these two checks? The method will return at L161 or L164 if any of them is false
```suggestion
  if (!_young_gen_has_alloc_failure) {
```